lane.py

```diff
@@ -387,897 +372,796 @@ def has_next_page(self):
+            value = tuple_to_numericrange(value)
+        self._target_age = value
+        if value.upper >= 18:
+            audiences.append(Classifier.AUDIENCE_ADULT)
+        if value.lower < Classifier.YOUNG_ADULT_AGE_CUTOFF:
+            audiences.append(Classifier.AUDIENCE_CHILDREN)
+        if value.upper >= Classifier.YOUNG_ADULT_AGE_CUTOFF:
+            audiences.append(Classifier.AUDIENCE_YOUNG_ADULT)
+        self._audiences = audiences
+
+    @hybrid_property
+    def list_datasource(self):
+        return self._list_datasource
+
+    @list_datasource.setter
+    def set_list_datasource(self, value):
+        """Setting .list_datasource to a non-null value wipes out any specific
+        CustomLists previously associated with this Lane.
+        """
+        if value:
+            self.customlists = []
+            value = value.id
+        self._list_datasource_id = value
 
-        distinct = False
-        if self.list_data_source_id or self.list_ids:
-            # One book can show up on more than one list; we need to
-            # add a DISTINCT clause.
-            distinct = True
+    @property
+    def uses_customlists(self):
+        """Does the works() implementation for this Lane look for works on
+        CustomLists?
+        """
+        if self.customlists or self.list_datasource:
+            return True
+        if (self.parent and self.inherit_parent_restrictions 
+            and self.parent.uses_customlists):
+            return True
+        return False        
 
-            if work_model == Work:
-                clause = CustomListEntry.work_id==work_model.id
-            else:
-                clause = CustomListEntry.work_id==work_model.works_id
-            q = q.join(CustomListEntry, clause)
-            if self.list_data_source_id:
-                q = q.join(CustomListEntry.customlist).filter(
-                    CustomList.data_source_id==self.list_data_source_id)
-            else:
-                q = q.filter(
-                    CustomListEntry.list_id.in_(self.list_ids)
-                )
-            if self.list_seen_in_previous_days:
-                cutoff = datetime.datetime.utcnow() - datetime.timedelta(
-                    self.list_seen_in_previous_days
-                )
-                q = q.filter(CustomListEntry.most_recent_appearance
-                             >=cutoff)
+    @property
+    def genre_ids(self):
+        """Find the database ID of every Genre such that a Work classified in
+        that Genre should be in this Lane.
 
-        if facets:
-            q = facets.apply(self._db, q, work_model, edition_model,
-                             distinct=distinct)
-        if pagination:
-            q = pagination.apply(q)
+        :return: A list of genre IDs, or None if this Lane does not
+        consider genres at all.
+        """
+        if not hasattr(self, '_genre_ids'):
+            self._genre_ids = self._gather_genre_ids()
+        return self._genre_ids
```

Comment:
I'm concerned that a change to the WorkGenres for a lane doesn't force this to be recalculated. I'll probably need to add an event listener.